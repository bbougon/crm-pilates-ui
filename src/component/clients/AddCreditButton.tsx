import { Grid, IconButton } from "@mui/material";
import AddBoxIcon from "@mui/icons-material/AddBox";
import * as React from "react";

type AddCreditButtonProps = {
  onAddCreditButton: () => void;
  disabled?: boolean;
};
export const AddCreditButton = ({
  onAddCreditButton,
  disabled,
}: AddCreditButtonProps) => {
  return (
    <Grid
      container
      direction="row"
      sx={{
        paddingBottom: "4px",
        alignItems: "center",
      }}
    >
      <Grid
        item
        xs={12}
        sx={{
          display: "flex",
          justifyContent: "flex-end",
        }}
      >
        <IconButton
          aria-label="add"
          disabled={disabled}
          onClick={onAddCreditButton}
        >
          <AddBoxIcon />
        </IconButton>
      </Grid>
    </Grid>
  );
};
