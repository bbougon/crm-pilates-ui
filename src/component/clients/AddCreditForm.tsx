import * as React from "react";
import {BaseSyntheticEvent, MouseEvent, useState} from "react";
import {Subjects} from "../../features/domain/subjects";
import {FormControl, Grid, InputLabel, MenuItem, Select, SelectChangeEvent, TextField} from "@mui/material";
import {Button} from "@material-ui/core";

type AddCreditFormProps = {
    subjects: {subject: Subjects, title: string}[] | []
    onAddCredits: (creditsAmount: number, subject: Subjects) => void
}
export const AddCreditForm = ({onAddCredits, subjects}: AddCreditFormProps)  => {

    const [subject, setSubject] = useState<Subjects | "">("")
    const [creditsAmount, setCreditsAmount] = useState<number>(0)

    const onSubjectChanged = (e: SelectChangeEvent<Subjects | unknown>) => setSubject(e.target.value as Subjects)
    const onCreditsAmountChanged = (e: BaseSyntheticEvent) => setCreditsAmount(e.target.value)
    const onSubmitClicked = (_: MouseEvent) => {
        const value: number = +creditsAmount
        onAddCredits(value, subject as Subjects)
    }

    return (
        <Grid container direction="row" sx={{
            paddingBottom: "4px",
            alignItems: "center"
        }}>
            <Grid item xs={2} sx={{
                display: 'flex',
                textAlign: 'left',
                justifyContent: 'flex-start',
            }}>
                <FormControl fullWidth>
                    <InputLabel id="subject-select-label">Subject</InputLabel>
                    <Select
                        labelId="subject-select-label"
                        id="subject-select"
                        value={subject || ""}
                        required
                        placeholder="Select a subject"
                        label="Subject"
                        variant="standard"
                        onChange={onSubjectChanged}
                        size="small"
                    >
                        {subjects.map(subject => <MenuItem key={subject.subject}
                                                                    value={subject.subject}>{subject.title}</MenuItem>)}
                    </Select>
                </FormControl>
            </Grid>
            <Grid item xs={4} sx={{
                display: 'flex',
                justifyContent: 'flex-start',
                paddingLeft: '16px',
            }}
            >
                <FormControl>
                    <TextField id="credits-amount"
                               size="small"
                               type="number"
                               label="Amount of credits"
                               required
                               variant="standard"
                               onChange={onCreditsAmountChanged}
                               value={creditsAmount || ""}
                               aria-describedby="credits-amount-help"/>
                </FormControl>
            </Grid>
            <Grid item xs={5} sx={{
                display: 'flex',
                justifyContent: 'flex-start',
            }}>
                <Button size="small" onClick={onSubmitClicked}>Add credits</Button>
            </Grid>
        </Grid>
    )
}