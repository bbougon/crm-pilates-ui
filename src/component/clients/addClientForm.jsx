import {
    Accordion,
    AccordionActions,
    AccordionDetails,
    AccordionSummary,
    Button,
    Divider,
    FormControl,
    Grid,
    TextField,
    Typography
} from "@material-ui/core";
import * as React from "react";
import {useState} from "react";
import ExpandMoreIcon from "@material-ui/icons/ExpandMore";
import styled from "styled-components";
import {useDispatch, useSelector} from "react-redux";
import {clientStatuses, createClient} from "../../features/clientsSlice";
import {DisplayError} from "../errors/displayError";


const Wrapper = styled.div`
    display: flex;
    flex-direction: column;
    flex: auto;
    width: 100%;
`;


function ClientAccordionSummary() {
    return (
        <AccordionSummary
            expandIcon={<ExpandMoreIcon/>}
            aria-controls="panel1c-content"
            id="panel1c-header"
            sx={{textAlign: "center"}}>
            <Typography variant="h6" color="textSecondary">Add a new client</Typography>
        </AccordionSummary>
    )
}

export const AddClientForm = () => {

    const dispatch = useDispatch()

    const error = useSelector((state => state.clients.error))
    const status = useSelector((state => state.clients.status))
    let errorContent = undefined

    const [firstname, setFirstname] = useState('')
    const [lastname, setLastname] = useState('')

    const onFirstnameChanged = (e) => setFirstname(e.target.value)
    const onLastnameChanged = (e) => setLastname(e.target.value)

    const onSubmitClicked = async () => {
        await dispatch(createClient({firstname, lastname}))
        setLastname('')
        setFirstname('')
    }

    if (status === clientStatuses.CREATION_FAILED) {
        errorContent = (
            <Grid container>
                <Grid item xs={12} md={12}>
                    <DisplayError error={error}/>
                </Grid>
            </Grid>
        )
    }

    return (
        <Accordion>
            <ClientAccordionSummary/>
            <AccordionDetails>
                <Wrapper>
                    <Grid container>
                        <Grid item xs={12} md={3}>
                            <FormControl>
                                <TextField id="client-name" className="sizeSmall"
                                           label="Client's name"
                                           helperText="Provide a client's name"
                                           required
                                           onChange={onLastnameChanged}
                                           aria-describedby="client-name-help"
                                           value={lastname}
                                />
                            </FormControl>
                        </Grid>
                    </Grid>
                    <Grid container>
                        <Grid item xs={12} md={3}>
                            <FormControl>
                                <TextField id="client-first-name" className="sizeSmall"
                                           label="Client's firstname"
                                           helperText="Provide a client's firstname"
                                           required
                                           onChange={onFirstnameChanged}
                                           aria-describedby="client-first-name-help"
                                           value={firstname}
                                />
                            </FormControl>
                        </Grid>
                    </Grid>
                    {errorContent}
                </Wrapper>
            </AccordionDetails>
            <Divider/>
            <AccordionActions>
                <Button onClick={onSubmitClicked} disabled={firstname === "" || lastname === ""}>Submit</Button>
            </AccordionActions>
        </Accordion>
    );

}
