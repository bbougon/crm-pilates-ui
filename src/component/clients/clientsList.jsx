import * as React from "react";
import {useEffect} from "react";
import {Card, Grid, Typography} from "@material-ui/core";
import {useDispatch, useSelector} from "react-redux";
import {fetchClients, selectAllClients} from "../../features/clientsSlice";
import {Avatar, CardContent, Stack} from "@mui/material";
import PersonIcon from '@mui/icons-material/Person';
import {DisplayError} from "../errors/displayError";


const Client = ({client}) => {
    return (
        <Grid item xs={12}>
            <Card>
                <CardContent sx={{textAlign: "left"}}>
                    <Stack direction="row" spacing={2} sx={{alignItems: "center"}}>
                        <Avatar sx={{width: 24, height: 24}}>
                            <PersonIcon/>
                        </Avatar>
                        <Typography variant="h6" gutterBottom>
                            {client.lastname}
                        </Typography>
                        <Typography gutterBottom>
                            {client.firstname}
                        </Typography>
                    </Stack>
                </CardContent>
            </Card>
        </Grid>
    )
};

export const ClientsList = () => {

    const dispatch = useDispatch();
    const clients = useSelector(selectAllClients);
    const error = useSelector((state => state.clients.error))
    const status = useSelector((state => state.clients.status))
    let content

    useEffect(() => {
        dispatch(fetchClients())
    }, [dispatch])

    content = clients.map((client) => (<Client key={client.id} client={client}/>))
    if (status === "failed") {
        content = <DisplayError error={error}/>
    }

    return (
        <Grid container spacing={1}>
            {content}
        </Grid>
    )

}
