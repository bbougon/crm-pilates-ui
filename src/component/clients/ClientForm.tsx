import {
    Accordion,
    AccordionActions,
    AccordionDetails,
    AccordionSummary,
    Button,
    Divider,
    FormControl,
    TextField,
    Typography
} from "@material-ui/core";
import * as React from "react";
import {useState} from "react";
import ExpandMoreIcon from "@material-ui/icons/ExpandMore";
import styled from "styled-components";
import {useDispatch, useSelector} from "react-redux";
import {ClientStatus, createClient} from "../../features/clientsSlice";
import {DisplayError} from "../errors/DisplayError";
import {ErrorMessage} from "../../features/errors";
import {RootState} from "../../app/store";
import {Subjects} from "../../features/domain/subjects";
import {AddCreditButton} from "./AddCreditButton";
import {AddCreditForm} from "./AddCreditForm";
import {subjects} from "../../utils/translation";
import {CreditItem} from "./CreditItem";
import {Grid} from "@mui/material";


const Wrapper = styled.div`
    display: flex;
    flex-direction: column;
    flex: auto;
    text-align: left;
    width: 100%;
`;

const AddClientAccordionSummary = () => {
    const props = {
        expandIcon: <ExpandMoreIcon/>,
        id: "panel1c-header"
    }
    return (
        <AccordionSummary {...props} aria-controls={"panel1c-content"}>
            <Typography variant="h6" color="textSecondary">Add a new client</Typography>
        </AccordionSummary>
    )
}

export const AddClientForm = () => {

    const dispatch = useDispatch()

    const errorMessages: ErrorMessage[] = useSelector<RootState, ErrorMessage[]>((state => state.clients.error))
    const status: ClientStatus = useSelector<RootState, ClientStatus>((state => state.clients.status))
    const [addCreditForms, setAddCreditForms] = useState<any>(undefined)
    const [creditItems, setCreditItems] = useState<any[]>([])
    const [availableSubjects, setAvailableSubjects] = useState<{subject: Subjects, title: string}[]>(subjects)
    const [addCreditsInProgress, setAddCreditsInProgress] = useState(false)
    let errorContent = undefined

    const [firstname, setFirstname] = useState('')
    const [lastname, setLastname] = useState('')
    const [credits, setCredits] = useState<{ value: number, subject: string }[] | []>([])

    const onFirstnameChanged = (e: any) => setFirstname(e.target.value)
    const onLastnameChanged = (e: any) => setLastname(e.target.value)

    const onSubmitClicked = async () => {
        dispatch(createClient({firstname, lastname, credits}))
        setLastname('')
        setFirstname('')
        setCreditItems([])
    }

    const onAddCredits = (creditsAmount: number, subject: Subjects) => {
        setCredits([...credits, {value: creditsAmount as number, subject: subject}])
        setCreditItems([...creditItems,
            <Grid key={"credit-item-container-".concat(Math.random().toString())} container direction="row" sx={{
                paddingTop: '4px'
            }}>
                <CreditItem key={"credit-item".concat(Math.random().toString())}
                                                        credits={{value: creditsAmount, subject: subject}}/>
            </Grid>])
        setAddCreditForms(undefined)
        setAddCreditsInProgress(false)
        setAvailableSubjects(availableSubjects.filter(currentSubject => subject !== currentSubject.subject))
    }

    const onAddCreditButton = () => {
        setAddCreditForms(<AddCreditForm key={`add-credit-form-`.concat(Math.random().toString())} subjects={availableSubjects}
                                         onAddCredits={onAddCredits}/>)
        setAddCreditsInProgress(true)
    }

    if (status === ClientStatus.CREATION_FAILED) {
        errorContent = (
            <Grid container>
                <Grid item xs={12} md={12}>
                    <DisplayError {...{error: errorMessages}}/>
                </Grid>
            </Grid>
        )
    }

    return (
        <Accordion>
            <AddClientAccordionSummary/>
            <AccordionDetails>
                <Wrapper>
                    <Grid container>
                        <Grid item xs={12} md={3}>
                            <FormControl>
                                <TextField id="client-name" className="sizeSmall"
                                           label="Client's name"
                                           helperText="Provide a client's name"
                                           required
                                           onChange={onLastnameChanged}
                                           aria-describedby="client-name-help"
                                           value={lastname}
                                />
                            </FormControl>
                        </Grid>
                        <Grid item xs={12} md={3}>
                            <FormControl>
                                <TextField id="client-first-name" className="sizeSmall"
                                           label="Client's firstname"
                                           helperText="Provide a client's firstname"
                                           required
                                           onChange={onFirstnameChanged}
                                           aria-describedby="client-first-name-help"
                                           value={firstname}
                                />
                            </FormControl>
                        </Grid>
                    </Grid>
                    <Grid key={"credits-items-".concat(Math.random().toString())} container direction="column">
                        {creditItems}
                    </Grid>
                    {addCreditForms}
                    <AddCreditButton key={`add-credit-buton-`.concat(Math.random().toString())}
                                     disabled={addCreditsInProgress || availableSubjects.length === 0} onAddCreditButton={onAddCreditButton}/>
                    {errorContent}
                </Wrapper>
            </AccordionDetails>
            <Divider/>
            <AccordionActions>
                <Button onClick={onSubmitClicked} disabled={firstname === "" || lastname === ""}>Submit</Button>
            </AccordionActions>
        </Accordion>
    );

}
