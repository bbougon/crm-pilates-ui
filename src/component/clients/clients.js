import * as React from "react";
import {MainContainer} from "../../const/containers";
import {Grid} from "@material-ui/core";
import {CreateClientForm} from "./createClientForm";
import {ClientsList} from "./clientsList";


export class Clients extends React.Component {


    render() {
        return (
            <MainContainer>
                <Grid container>
                    <Grid item xs={11}>
                        <>
                            <CreateClientForm />
                        </>
                    </Grid>
                    <Grid item xs={11}>
                        <>
                            <ClientsList />
                        </>
                    </Grid>
                </Grid>
            </MainContainer>
        );
    }
}

export default Clients;