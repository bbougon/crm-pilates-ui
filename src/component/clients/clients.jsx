import * as React from "react";
import {MainContainer} from "../../const/containers";
import {Grid} from "@material-ui/core";
import {AddClientForm} from "./addClientForm";
import {ClientsList} from "./clientsList";


export class Clients extends React.Component {


    render() {
        return (
            <MainContainer>
                <Grid container>
                    <Grid item xs={11}>
                        <>
                            <AddClientForm />
                        </>
                    </Grid>
                    <Grid item xs={11}>
                        <>
                            <ClientsList />
                        </>
                    </Grid>
                </Grid>
            </MainContainer>
        );
    }
}

export default Clients;