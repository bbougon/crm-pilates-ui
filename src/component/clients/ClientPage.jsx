import * as React from "react";
import {Grid} from "@material-ui/core";
import {ClientsList} from "./ClientsList";
import {AddClientForm} from "./ClientForm";
import {MainContainer} from "../../const/containers";


export class Clients extends React.Component {


    render() {
        return (
            <MainContainer>
                <Grid container>
                    <Grid item xs={11}>
                        <>
                            <AddClientForm />
                        </>
                    </Grid>
                    <Grid item xs={11}>
                        <>
                            <ClientsList />
                        </>
                    </Grid>
                </Grid>
            </MainContainer>
        );
    }
}

export default Clients;