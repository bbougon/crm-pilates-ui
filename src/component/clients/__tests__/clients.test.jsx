import React from "react";
import {Clients} from "../clients";
import {prettyDOM, screen, waitFor} from '@testing-library/react';
import userEvent from "@testing-library/user-event";
import {rest} from 'msw'
import {setupServer} from 'msw/node'
import {render} from "../../../test-utils/test-utils";
import {client, ClientsBuilder} from "../../../test-utils/clients/clients";
import {APIDetail, APIErrorBody, ServerBuilder} from "../../../test-utils/server/server";


describe('Clients page', function () {
    const server = new ServerBuilder("/clients", "get").withBody([]).build()

    beforeAll(() => server.listen())

    afterEach(() => server.resetHandlers())

    afterAll(() => server.close())

    it('should be displayed', () => {
        const {container} = render(<Clients/>)

        expect(prettyDOM(container)).toMatchSnapshot()
    })

    describe('fetches clients when loading', function () {
        describe("retrieve them", () => {
            const server = new ServerBuilder("/clients")
                .withBody(new ClientsBuilder()
                    .withClient(client())
                    .withClient(client("Pierre", "Martin", "33da6f24-efda-4c16-b8af-e5e822fc5860"))
                    .withClient(client("Henri", "Verneuil", "33da6bca-efda-4c16-b8af-e5e822fc5901"))
                    .build())
                .build()

            beforeAll(() => server.listen())

            afterEach(() => server.resetHandlers())

            afterAll(() => server.close())

            it('and should display them', async () => {
                render(<Clients/>)

                expect(await screen.findByText("Doe", {selector: 'h6'})).toBeInTheDocument()
                expect(await screen.findByText("John")).toBeInTheDocument()
                expect(await screen.findByText("Martin", {selector: 'h6'})).toBeInTheDocument()
                expect(await screen.findByText("Pierre")).toBeInTheDocument()
                expect(await screen.findByText("Verneuil", {selector: 'h6'})).toBeInTheDocument()
                expect(await screen.findByText("Henri")).toBeInTheDocument()
            })
        })

        describe("faces an error", () => {
            const server = new ServerBuilder("/clients")
                .withStatus(422)
                .withBody(new APIErrorBody()
                    .dummyDetail()
                    .build())
                .build()

            beforeAll(() => server.listen())

            afterEach(() => server.resetHandlers())

            afterAll(() => server.close())

            it("and should display it", async () => {
                render(<Clients/>)

                expect(await screen.findByText("An error occurred (see message below):", {selector: 'h5'})).toBeInTheDocument()
                expect(await screen.findByText("an error message", {selector: 'p'})).toBeInTheDocument()
                expect(await screen.findByText("an error type", {selector: 'p'})).toBeInTheDocument()
            })
        })

    })

    describe('displays a form to create a client', function () {
        const server = setupServer(
            rest.get('/clients', (req, res, ctx) => {
                return res(ctx.json([]))
            }),
            rest.post('/clients', (req, res, ctx) => {
                return res(ctx.json(client()))
            }),
        )

        beforeAll(() => server.listen())

        afterEach(() => server.resetHandlers())

        afterAll(() => server.close())

        it('and should create John Doe', async () => {
            render(<Clients/>)

            userEvent.click(screen.getByRole("button", {name: /add a new client/i}))
            userEvent.type(screen.getByText("Client's name"), "Doe")
            userEvent.type(screen.getByText("Client's firstname"), "John")
            userEvent.click(screen.getByRole("button", {name: /submit/i}))

            expect(await screen.queryByDisplayValue("Doe")).toBeInTheDocument()
            expect(await screen.queryByDisplayValue("John")).toBeInTheDocument()
            await waitFor(() => expect(screen.getByLabelText("Client's name *", {selector: 'input'})).toHaveValue(""))
            await waitFor(() => expect(screen.getByLabelText("Client's firstname *", {selector: 'input'})).toHaveValue(""))
        })

        describe("faces an error when creating a client", function() {
            const server = setupServer(
                rest.get('/clients', (req, res, ctx) => {
                    return res(ctx.json([]))
                }),
                rest.post('/clients', (req, res, ctx) => {
                    let body = new APIErrorBody()
                        .withDetail(APIDetail("You must provide the client firstname", "value_error"))
                        .withDetail(APIDetail("You must provide the client lastname", "value_error"))
                        .build();
                    return res(
                        ctx.status(422),
                        ctx.json(body)
                    )
                }),
            )

            beforeAll(() => server.listen())

            afterEach(() => server.resetHandlers())

            afterAll(() => server.close())

            it("should display the error", async () => {
                render(<Clients/>)

                userEvent.click(screen.getByRole("button", {name: /add a new client/i}))
                userEvent.type(screen.getByText("Client's name"), "{empty}")
                userEvent.type(screen.getByText("Client's firstname"), "{empty}")
                userEvent.click(screen.getByRole("button", {name: /submit/i}))

                expect(await screen.findByText("An error occurred (see message below):", {selector: 'h5'})).toBeInTheDocument()
                await waitFor(() => expect(screen.queryByText("You must provide the client firstname", {selector: 'p'})).toBeInTheDocument())
                await waitFor(() => expect(screen.queryByText("You must provide the client lastname", {selector: 'p'})).toBeInTheDocument())
                await waitFor(() => expect(screen.queryAllByText("value_error", {selector: 'p'})).toBeTruthy())
            })
        })
    })
})