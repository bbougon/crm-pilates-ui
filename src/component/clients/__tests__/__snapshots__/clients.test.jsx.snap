// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clients should display the client page 1`] = `
"[36m<div>[39m
  [36m<div[39m
    [33mclass[39m=[32m\\"sc-bdfBQB gOWVrL\\"[39m
  [36m>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"MuiGrid-root MuiGrid-container\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-11\\"[39m
      [36m>[39m
        [36m<div[39m
          [33mclass[39m=[32m\\"MuiPaper-root MuiAccordion-root MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded\\"[39m
        [36m>[39m
          [36m<div[39m
            [33maria-controls[39m=[32m\\"panel1c-content\\"[39m
            [33maria-disabled[39m=[32m\\"false\\"[39m
            [33maria-expanded[39m=[32m\\"false\\"[39m
            [33mclass[39m=[32m\\"MuiButtonBase-root MuiAccordionSummary-root\\"[39m
            [33mid[39m=[32m\\"panel1c-header\\"[39m
            [33mrole[39m=[32m\\"button\\"[39m
            [33msx[39m=[32m\\"[object Object]\\"[39m
            [33mtabindex[39m=[32m\\"0\\"[39m
          [36m>[39m
            [36m<div[39m
              [33mclass[39m=[32m\\"MuiAccordionSummary-content\\"[39m
            [36m>[39m
              [36m<h6[39m
                [33mclass[39m=[32m\\"MuiTypography-root MuiTypography-h6 MuiTypography-colorTextSecondary\\"[39m
              [36m>[39m
                [0mAdd a new client[0m
              [36m</h6>[39m
            [36m</div>[39m
            [36m<div[39m
              [33maria-disabled[39m=[32m\\"false\\"[39m
              [33maria-hidden[39m=[32m\\"true\\"[39m
              [33mclass[39m=[32m\\"MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon MuiIconButton-edgeEnd\\"[39m
            [36m>[39m
              [36m<span[39m
                [33mclass[39m=[32m\\"MuiIconButton-label\\"[39m
              [36m>[39m
                [36m<svg[39m
                  [33maria-hidden[39m=[32m\\"true\\"[39m
                  [33mclass[39m=[32m\\"MuiSvgIcon-root\\"[39m
                  [33mfocusable[39m=[32m\\"false\\"[39m
                  [33mviewBox[39m=[32m\\"0 0 24 24\\"[39m
                [36m>[39m
                  [36m<path[39m
                    [33md[39m=[32m\\"M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z\\"[39m
                  [36m/>[39m
                [36m</svg>[39m
              [36m</span>[39m
              [36m<span[39m
                [33mclass[39m=[32m\\"MuiTouchRipple-root\\"[39m
              [36m/>[39m
            [36m</div>[39m
          [36m</div>[39m
          [36m<div[39m
            [33mclass[39m=[32m\\"MuiCollapse-root MuiCollapse-hidden\\"[39m
            [33mstyle[39m=[32m\\"min-height: 0px;\\"[39m
          [36m>[39m
            [36m<div[39m
              [33mclass[39m=[32m\\"MuiCollapse-wrapper\\"[39m
            [36m>[39m
              [36m<div[39m
                [33mclass[39m=[32m\\"MuiCollapse-wrapperInner\\"[39m
              [36m>[39m
                [36m<div[39m
                  [33mrole[39m=[32m\\"region\\"[39m
                [36m>[39m
                  [36m<div[39m
                    [33mclass[39m=[32m\\"MuiAccordionDetails-root\\"[39m
                  [36m>[39m
                    [36m<div[39m
                      [33mclass[39m=[32m\\"sc-gsTEea iKEupo\\"[39m
                    [36m>[39m
                      [36m<div[39m
                        [33mclass[39m=[32m\\"MuiGrid-root MuiGrid-container\\"[39m
                      [36m>[39m
                        [36m<div[39m
                          [33mclass[39m=[32m\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3\\"[39m
                        [36m>[39m
                          [36m<div[39m
                            [33mclass[39m=[32m\\"MuiFormControl-root\\"[39m
                          [36m>[39m
                            [36m<div[39m
                              [33maria-describedby[39m=[32m\\"client-name-help\\"[39m
                              [33mclass[39m=[32m\\"MuiFormControl-root MuiTextField-root sizeSmall\\"[39m
                            [36m>[39m
                              [36m<label[39m
                                [33mclass[39m=[32m\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required\\"[39m
                                [33mdata-shrink[39m=[32m\\"false\\"[39m
                                [33mfor[39m=[32m\\"client-name\\"[39m
                                [33mid[39m=[32m\\"client-name-label\\"[39m
                              [36m>[39m
                                [0mClient's name[0m
                                [36m<span[39m
                                  [33maria-hidden[39m=[32m\\"true\\"[39m
                                  [33mclass[39m=[32m\\"MuiFormLabel-asterisk MuiInputLabel-asterisk\\"[39m
                                [36m>[39m
                                  [0mâ€‰[0m
                                  [0m*[0m
                                [36m</span>[39m
                              [36m</label>[39m
                              [36m<div[39m
                                [33mclass[39m=[32m\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl\\"[39m
                              [36m>[39m
                                [36m<input[39m
                                  [33maria-describedby[39m=[32m\\"client-name-helper-text\\"[39m
                                  [33maria-invalid[39m=[32m\\"false\\"[39m
                                  [33mclass[39m=[32m\\"MuiInputBase-input MuiInput-input\\"[39m
                                  [33mid[39m=[32m\\"client-name\\"[39m
                                  [33mrequired[39m=[32m\\"\\"[39m
                                  [33mtype[39m=[32m\\"text\\"[39m
                                  [33mvalue[39m=[32m\\"\\"[39m
                                [36m/>[39m
                              [36m</div>[39m
                              [36m<p[39m
                                [33mclass[39m=[32m\\"MuiFormHelperText-root Mui-required\\"[39m
                                [33mid[39m=[32m\\"client-name-helper-text\\"[39m
                              [36m>[39m
                                [0mProvide a client's name[0m
                              [36m</p>[39m
                            [36m</div>[39m
                          [36m</div>[39m
                        [36m</div>[39m
                      [36m</div>[39m
                      [36m<div[39m
                        [33mclass[39m=[32m\\"MuiGrid-root MuiGrid-container\\"[39m
                      [36m>[39m
                        [36m<div[39m
                          [33mclass[39m=[32m\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-3\\"[39m
                        [36m>[39m
                          [36m<div[39m
                            [33mclass[39m=[32m\\"MuiFormControl-root\\"[39m
                          [36m>[39m
                            [36m<div[39m
                              [33maria-describedby[39m=[32m\\"client-first-name-help\\"[39m
                              [33mclass[39m=[32m\\"MuiFormControl-root MuiTextField-root sizeSmall\\"[39m
                            [36m>[39m
                              [36m<label[39m
                                [33mclass[39m=[32m\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required\\"[39m
                                [33mdata-shrink[39m=[32m\\"false\\"[39m
                                [33mfor[39m=[32m\\"client-first-name\\"[39m
                                [33mid[39m=[32m\\"client-first-name-label\\"[39m
                              [36m>[39m
                                [0mClient's firstname[0m
                                [36m<span[39m
                                  [33maria-hidden[39m=[32m\\"true\\"[39m
                                  [33mclass[39m=[32m\\"MuiFormLabel-asterisk MuiInputLabel-asterisk\\"[39m
                                [36m>[39m
                                  [0mâ€‰[0m
                                  [0m*[0m
                                [36m</span>[39m
                              [36m</label>[39m
                              [36m<div[39m
                                [33mclass[39m=[32m\\"MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl\\"[39m
                              [36m>[39m
                                [36m<input[39m
                                  [33maria-describedby[39m=[32m\\"client-first-name-helper-text\\"[39m
                                  [33maria-invalid[39m=[32m\\"false\\"[39m
                                  [33mclass[39m=[32m\\"MuiInputBase-input MuiInput-input\\"[39m
                                  [33mid[39m=[32m\\"client-first-name\\"[39m
                                  [33mrequired[39m=[32m\\"\\"[39m
                                  [33mtype[39m=[32m\\"text\\"[39m
                                  [33mvalue[39m=[32m\\"\\"[39m
                                [36m/>[39m
                              [36m</div>[39m
                              [36m<p[39m
                                [33mclass[39m=[32m\\"MuiFormHelperText-root Mui-required\\"[39m
                                [33mid[39m=[32m\\"client-first-name-helper-text\\"[39m
                              [36m>[39m
                                [0mProvide a client's firstname[0m
                              [36m</p>[39m
                            [36m</div>[39m
                          [36m</div>[39m
                        [36m</div>[39m
                      [36m</div>[39m
                    [36m</div>[39m
                  [36m</div>[39m
                  [36m<hr[39m
                    [33mclass[39m=[32m\\"MuiDivider-root\\"[39m
                  [36m/>[39m
                  [36m<div[39m
                    [33mclass[39m=[32m\\"MuiAccordionActions-root MuiAccordionActions-spacing\\"[39m
                  [36m>[39m
                    [36m<button[39m
                      [33mclass[39m=[32m\\"MuiButtonBase-root MuiButton-root MuiButton-text Mui-disabled Mui-disabled\\"[39m
                      [33mdisabled[39m=[32m\\"\\"[39m
                      [33mtabindex[39m=[32m\\"-1\\"[39m
                      [33mtype[39m=[32m\\"button\\"[39m
                    [36m>[39m
                      [36m<span[39m
                        [33mclass[39m=[32m\\"MuiButton-label\\"[39m
                      [36m>[39m
                        [0mSubmit[0m
                      [36m</span>[39m
                    [36m</button>[39m
                  [36m</div>[39m
                [36m</div>[39m
              [36m</div>[39m
            [36m</div>[39m
          [36m</div>[39m
        [36m</div>[39m
      [36m</div>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-11\\"[39m
      [36m>[39m
        [36m<div[39m
          [33mclass[39m=[32m\\"MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1\\"[39m
        [36m/>[39m
      [36m</div>[39m
    [36m</div>[39m
  [36m</div>[39m
[36m</div>[39m"
`;
